File: src/main/java/com/tms/csp/transforms/FlattenOrs.java
package com.tms.csp.transforms;

import com.tms.csp.ast.Exp;
import com.tms.csp.ast.Or;

import static com.google.common.base.Preconditions.checkArgument;

public class FlattenOrs extends BaseTransformer {

    @Override
    protected Exp local(Exp in) {
        return flattenOr(in.asOr());
    }

    @Override
    protected boolean executeLocal(Exp in) {
        if (in.isOr()) {
            for (Exp arg : in.argIt()) {
                if (arg.isOr()) {
                    return true;
                }
            }
        }
        return false;
    }

    private Exp flattenOr(Or in) {
        checkArgument(in.argList().size() >= 2);
        assert !in.isNestedOr() : in.toString();
        return in;

    }

}

File: src/main/java/com/smartsoft/csp/transforms/FlattenOrs.kt
package com.smartsoft.csp.transforms

import com.smartsoft.csp.ast.Exp
import com.smartsoft.csp.ast.Or

import com.google.common.base.Preconditions.checkArgument

class FlattenOrs : BaseTransformer() {

    override fun local(inExp: Exp): Exp? {
        return flattenOr(inExp.asOr)
    }

    override fun executeLocal(inExp: Exp): Boolean {
        if (inExp.isOr) {
            for (arg in inExp.argIt) {
                if (arg.isOr) {
                    return true
                }
            }
        }
        return false
    }

    private fun flattenOr(inExp: Or): Exp {
        checkArgument(inExp.argList().size >= 2)
        assert(!inExp.isNestedOr) { inExp.toString() }
        return inExp

    }

}
